protocols/bftcosi/bftcosi.go

```diff
@@ -228,125 +237,97 @@ func (bft *ProtocolBFTCoSi) RegisterOnSignatureDone(fn func(*BFTSignature)) {
 	bft.onSignatureDone = fn
 }
 
-// Shutdown will close the dispatch-method if the protocol is stopped
-// before the normal termination.
+// Shutdown closes all channels in case we're done
 func (bft *ProtocolBFTCoSi) Shutdown() error {
-	bft.doneProcessing <- true
+	defer func() {
+		// In case the channels were already closed
+		recover()
+	}()
+	bft.closingMutex.Lock()
+	bft.closing = true
+	bft.closingMutex.Unlock()
+	close(bft.announceChan)
+	close(bft.commitChan)
+	close(bft.challengePrepareChan)
+	close(bft.challengeCommitChan)
+	close(bft.responseChan)
 	return nil
 }
 
-// startAnnouncementPrepare create its announcement for the prepare round and
-// sends it down the tree.
-func (bft *ProtocolBFTCoSi) startAnnouncement(t RoundType) error {
-	a := &Announce{
-		TYPE: t,
-	}
-	log.Lvl4("RoundType:", t)
-	return bft.SendToChildrenInParallel(a)
-}
-
 // handleAnnouncement passes the announcement to the right CoSi struct.
-func (bft *ProtocolBFTCoSi) handleAnnouncement(ann Announce) error {
-	log.Lvl4(bft.Name(), "RoundType:", ann.TYPE)
+func (bft *ProtocolBFTCoSi) handleAnnouncement(msg announceChan) error {
+	ann := msg.Announce
+	log.Lvl3(bft.Name(), "RoundType:", ann.TYPE)
+	if bft.closing {
+		return errors.New("Closing")
+	}
 	if bft.IsLeaf() {
 		return bft.startCommitment(ann.TYPE)
 	}
 	return bft.SendToChildrenInParallel(&ann)
 }
 
-// startCommitment sends the first commitment to the parent node
-func (bft *ProtocolBFTCoSi) startCommitment(t RoundType) error {
-	cm := bft.getCosi(t).CreateCommitment(nil)
-	log.Lvl4(bft.Name(), "RoundType:", t)
-	return bft.SendToParent(&Commitment{TYPE: t, Commitment: cm})
-}
-
 // handleCommitment collects all commitments from children and passes them
 // to the parent or starts the challenge-round if it's the root.
-func (bft *ProtocolBFTCoSi) handleCommitment(comm Commitment) error {
-	log.Lvl4(bft.Name(), "RoundType:", comm.TYPE)
-
-	var commitment abstract.Point
-	// store it and check if we have enough commitments
+func (bft *ProtocolBFTCoSi) handleCommitment(msgs []commitChan) error {
```

Comment:
sometimes the `handle.*`-methods get called directly with a structure. So instead of having `handle` with receiving channels and `handle` with structures/slices, I prefer having only one signature-type.